// forest-express-mongoose/dist/services/projection-builder.js
var _classCallCheck = require("babel-runtime/helpers/classCallCheck");

var _createClass = require("babel-runtime/helpers/createClass");

var ProjectionBuilder =
/*#__PURE__*/
function () {
  function ProjectionBuilder(schema) {
    _classCallCheck(this, ProjectionBuilder);

    this.schemaSmartFields = schema && schema.fields && schema.fields.filter(function (field) {
      return field.get;
    }).map(function (field) {
      return field.field;
    });
  } // NOTICE: Convert a list of field names into a mongo $project structure.


  _createClass(ProjectionBuilder, [{
    key: "findRequestSmartField",
    // NOTICE: Perform the intersection between schema and request smart fields.
    value: function findRequestSmartField(requestFieldsNames) {
      if (this.schemaSmartFields && requestFieldsNames) {
        return this.schemaSmartFields.filter(function (fieldName) {
          return requestFieldsNames.includes(fieldName);
        });
      }

      return [];
    }
  }, {
    key: "getProjection",
    value: function getProjection(fieldNames) {
      var requestSmartFields = this.findRequestSmartField(fieldNames);
      if (requestSmartFields.length) return null;
      return ProjectionBuilder.convertToProjection(fieldNames);
    }
  }], [{
    key: "convertToProjection",
    value: function convertToProjection(fieldsNames) {
      if (fieldsNames) {
        var fieldsObject = fieldsNames.reduce(function (fields, fieldName) {
          fields[fieldName] = 1;
          return fields;
        }, {});
        return {
          $project: fieldsObject
        };
      }

      return null;
    }
  }]);

  return ProjectionBuilder;
}();

module.exports = ProjectionBuilder;
